### PR TITLE
dev/translation#39 - Fix `core:install --lang`. Realign boot plugins.

### PR DESCRIPTION
**Pull request by totten**
Overview
--------

Suppose you run a command like:

```
cv core:install --lang=fr_FR --cms-base-url=http://example.com/
```

This fixes that use-case by realigning some of the bootstrap steps under `setup/plugins/installDatabase/*`.

This is a variation inspired by @demeritcowboy's #17214.

Before
------

* The above `core:install` command crashes because the `paths` service is unavailable.

* The `civi.setup.installDatabase` begins with these two steps:

    1. `InstallSchema.civi-setup.php` manipulates some global options (to make certain services work), and it also installs the schema, and it also reboots the `Container`.
    2. `BootstrapCivi.civi-setup.php` tries to boot again - e.g. it loads `civicrm.settings.php` and calls `CRM_Core_Config::singleton($loadFromDB=TRUE)`. But this is suspect because it neglects to set `$force=TRUE`.

After
-----

* The above `core:install` command runs.

* The `civi.setup.installDatabase` begins with these *three* steps:

    1. `Preboot.civi-setup.php` starts the system in a *partially booted* mode (ie with `$loadFromDB==FALSE`).
    2. `InstallSchema.civi-setup.php` just installs the schema.
    3. `BootstrapCivi.civi-setup.php` switches the system to a fully booted mode (ie with `$loadFromDB==TRUE`).

The essence of the change is to move some parts of `InstallSchema` and `BootstrapCivi` to a plugin `Preboot`, creating a sharper distinction between partial/preboot mode and the full/standard/bootstrapped mode.

Technical details
-----------------

When working with `setup/plugins/**`, I find it helpful to use the `--debug-event` option to see the active plugins, e.g.

```
$ cv core:install --lang fr_FR --cms-base-url='http://dmaster.bknix:8001' --debug-event=
civi.setup.installDatabase

[Event] civi.setup.installDatabase
+-------+------------------------------------------------------------------------------------+
| Order | Callable                                                                           |
+-------+------------------------------------------------------------------------------------+
| #1    | closure(civicrm/setup/plugins/common/LogEvents.civi-setup.php@31)                  |
| #2    | InstallSchemaPlugin->installDatabase()                                             |
| #3    | closure(civicrm/setup/plugins/installDatabase/BootstrapCivi.civi-setup.php@13)     |
| #4    | closure(civicrm/setup/plugins/installDatabase/SetLanguage.civi-setup.php@13)       |
| #5    | closure(civicrm/setup/plugins/installDatabase/InstallComponents.civi-setup.php@13) |
| #6    | closure(civicrm/setup/plugins/installDatabase/InstallExtensions.civi-setup.php@13) |
...
+-------+------------------------------------------------------------------------------------+
```

I tested this on a D7 build. My dev/test loop was basically these steps:

```
cv core:uninstall -f
cv core:install --lang fr_FR --cms-base-url='http://dmaster.bknix:8001' -vvv
echo 'select name, label, value from civicrm_option_value where name like "Meet%"' | cv sql
```

If the install succeeded, then the SQL command would display localized
labels for "Meeting" ("Reunion").

### CODE REVIEW

**Comment by demeritcowboy** [comment]
Thanks I can take a look.

File: setup/plugins/installDatabase/BootstrapCivi.civi-setup.php [hunk 1]
```diff
@@ -2,7 +2,15 @@
 /**
  * @file
  *
- * Perform the first system bootstrap.
+ * Perform the full bootstrap.
+ *
+ * GOAL: All the standard services (database, DAOs, translations, settings, etc) should be loaded.
+ *
+ * MECHANICS: This is basically calls `CRM_Core_Config::singleton(TRUE,TRUE)`.
```
Comment (demeritcowboy) [inline setup/plugins/installDatabase/BootstrapCivi.civi-setup.php:9]: Grammar-police.

File: setup/plugins/installDatabase/Preboot.civi-setup.php [hunk 2]
```diff
@@ -0,0 +1,49 @@
+<?php
+/**
+ * @file
+ *
+ * Perform an initial, partial bootstrap.
+ *
+ * GOAL: This should provide sufficient services for `InstallSchema` to generate
+ * For example, `ts()` needs to be online..
```
Comment (demeritcowboy) [inline setup/plugins/installDatabase/Preboot.civi-setup.php:8]: Dot police.

**Comment by demeritcowboy** [comment]
* General standards
    * [`r-explain`] __PASS__
    * [`r-user`] __PASS__
    * [`r-doc`] __PASS__
    * [`r-run`] __PASS__
      * I used drupal 8 since while I care about drupal 7 to me this applies more to drupal 8 and you've already tested with drupal 7.
      * I also wanted to test using an alternate path for l10n since that seems to be the future with drupal 8/composer, so I also tested that (with the same quickie hack as in the other PR).
      * And I tested drupal 8 without --lang, just to see.
      * I also noticed the additional `civi.setup.checkRequirements` so I very briefly tested that.
* Developer standards
    * [`r-tech`] __PASS__
      * I like that this version only touches the install process and doesn't have to fiddle with CRM_Core_Config.
    * [`r-code`] __PASS__
    * [`r-maint`] __Undecided__
      * Were you thinking of adding something like this https://github.com/civicrm/cv/blob/v0.3.2/tests/Command/CoreLifecycleTest.php#L34 using the equivalent of that sql select you mention in the description here?
    * [`r-test`] __PASS__

File: setup/plugins/installDatabase/BootstrapCivi.civi-setup.php (hunk 1 above)
Comment (totten) [reply setup/plugins/installDatabase/BootstrapCivi.civi-setup.php]: A life of poor grammatical choices will put you behind 💈💈💈💈

**Comment by mattwire** [comment]
test this please

**Comment by seamuslee001** [comment]
Jenkins re test this please

**Comment by demeritcowboy** [comment]
jenkins test this please

**Comment by seamuslee001** [comment]
I'm going to merge this as it seems sensible to me

**Pull request merged**